---
title: "Quality Control Report"
output: 
  html_document:
    toc: true
    toc_float:
        collapsed: true
  theme: sandstone
  highlight: pygments
toc-title: "Table of contents"
params:
  #input 
  rval_sheet: NA
  rval_sheet_target: NA
  name_var: NA
  grouping_var: NA
  donor_var: NA
  
  #normalization 
  normalization_mode: NA
  dropsnps: NA
  dropcphs: NA
  dropsex: NA
  maf: NA
  probes: NA
  
  #DMP 
  limma_voi: NA
  limma_covar: NA
  limma_inter: NA
  limma_arrayweights: NA
  limma_ebayes_trend: NA
  limma_ebayes_robust: NA
  

  rval_design: NA
  rval_contrasts: NA
  rval_voi: NA
  rval_dendrogram: NULL
  
  #heatmap 
  min_deltabeta: NA
  max_fdr: NA
  max_pvalue: NA
  clusteralg: NA
  groups2plot: NA
  contrasts2plot: NA
  Colv: NA
  distance: NA
  scale: NA
  removebatch: NA
  
  #DMRs
  dmrs_contrasts: NULL
  dmrs_rval_dendrogram: NULL
  dmrs_min_deltabeta: NA
  dmrs_max_fdr: NA
  dmrs_max_pvalue: NA
  dmrs_clusteralg: NA
  dmrs_groups2plot: NA
  dmrs_contrasts2plot: NA
  dmrs_regions2plot: NA
  dmrs_Colv: NA
  dmrs_distance: NA
  dmrs_scale: NA
  dmrs_removebatch: NA
  table_dmrs: NA
  filteredmcsea2heatmap: NULL
  
  #plots

  plot_green_intensities: NA
  plot_red_intensities: NA
  plot_failed_probes: NA
  plot_densityplotraw: NA
  plot_densityplotraw_green: NA
  plot_densityplotraw_red: NA
  plot_densityplotraw_II: NA
  plot_densityplotraw_all: NA
  plot_densityplot: NA
  plot_densityplot_green: NA
  plot_densityplot_red: NA
  plot_densityplot_II: NA
  plot_densityplot_all: NA
  plot_pcaplot: NA
  plot_corrplot: NA
  plot_boxplotraw: NA
  plot_boxplot: NA
  plot_qcraw: NA
  plot_bisulfiterawII: NA
  plot_sexprediction: NA
  plot_snpheatmap: NA
  plot_plotSA: NA
  
  table_pcaplot: NA 
  table_corrplot: NA 
  data_sexprediction: NA
  
  table_dmps: NA
  filteredlist2heatmap: NULL
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{css, echo=FALSE}
div, code {overflow-x:auto !important}
```

## Data

The group variable selected was __`r params$grouping_var`__, and the donor variable was __`r params$donor_var`__ .

Finally, the selected samples added to the RGSet file were:

```{r sample_sheet, echo=FALSE, warning=FALSE, message = FALSE}
DT::datatable(params$rval_sheet_target, extensions = 'Buttons',
    style = "bootstrap",
    rownames = FALSE,
    selection = "single",
    options = list(dom = 'Blfrtip',
    columnDefs = list(list(
      targets = match("Basename",colnames(params$rval_sheet_target)) - 1, visible = FALSE)),
     #scrollX = TRUE,
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10,25,50,-1), c(10,25,50,"All"))))
```

The following samples were excluded for the analysis:

```{r samples_excluded, echo=FALSE, warning=FALSE, message = FALSE, comment=NA}

cat(params$rval_sheet[[params$name_var]][!(params$rval_sheet[[params$name_var]] %in% params$rval_sheet_target[[params$name_var]])], 
    sep = ", ")

```

***

## Intensities boxplots

__Green intensities__

```{r green_intensities, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_green_intensities
```

__Red intensities__

```{r red_intensities, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_red_intensities
```

<br>

## Failed probes

The probe failure rate is calculated for each sample and two thresholds are represented (5%, 10%).

```{r failed_probes, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_failed_probes
```

<br>

## Normalization options

The selected normalization method was __`r params$normalization_mode`__. In the next plots, we can see the comparison of the processed data with the raw data.

After Minfi normalization, we obtained a GenomicRatioSet with some transformations depending on the type of normalization chosen. To understand better the different classes of Minfi packages, and their relations depending of the normalization method, please read this [vignette](https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html) of Minfi creators (it is also valid for Illumina EPIC arrays).

Options selected:

* Drop SNPs: __`r params$dropsnps`__ *`r if(params$dropsnps) paste('minimum MAF filtered:', params$maf)`*
* Drop CpHs: __`r params$dropcphs`__
* Drop X/Y Chromosomes: __`r params$dropsex`__

After normalization, __`r params$probes`__ DNA methylation positions remained.

<br>

### Density plot  

__Raw__

```{r density_plot_raw_Igreen, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_densityplotraw_all
```

__Processed `r params$normalization_mode`__

```{r density_plot_processed_Igreen, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_densityplot_all
```

<br>

## SNPs Heatmap

The methylation array has 65 specific SNP probes. These SNP probes are intended to be used for sample tracking and sample mixups. Each SNP probe ought to have values clustered around 3 distinct values corresponding to homo-, and hetero-zygotes. Therefore, different samples of the same donor should cluster together.

__Processed `r params$normalization_mode`__

```{r snps_heatmap, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_snpheatmap
```

<br>

## Sex Prediction

Depending on the average chromosome Y signal and the average chromosome X signal is possible to predict the sex of the sample donors.

```{r sex_prediction_plot, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_sexprediction
```

```{r sex_prediction_table, echo=FALSE, warning=FALSE, message = FALSE}
knitr::kable(data.frame(name=params$rval_sheet_target[[params$name_var]],sex=params$data_sexprediction))
```

<br>

## Batch effects

Correlating principal components with variables we can determine if Beta values are related to our variable of interest or other variables. This can also be useful to determine possible errors in the sample hybridization randomization and to select covariables to add to the linear model.


```{r corrplot, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_corrplot
```

Not useful variables are discarded and the variable type is autodetected. The autodetected variable types were:

```{r corrtable, echo=FALSE, warning=FALSE, message = FALSE}
DT::datatable(params$table_corrplot, extensions = 'Buttons',
    style = 'bootstrap',
    rownames = FALSE,
    selection = "single",
    options = list(dom = 'Blfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10,25,50,-1),
    c(10,25,50,"All"))))
```

***

## System and analysis information

This report was generated at __`r Sys.time()`__.

The session information was the following:

```{r system_info, echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
print(sessionInfo())
```






