---
title: "Complete Report"
output: 
  html_document:
    toc: true
    toc_float:
        collapsed: true
  theme: sandstone
  highlight: pygments
toc-title: "Table of contents"
params:
  #input 
  rval_sheet: NA
  rval_sheet_target: NA
  name_var: NA
  grouping_var: NA
  donor_var: NA
  
  #normalization 
  normalization_mode: NA
  dropsnps: NA
  dropcphs: NA
  dropsex: NA
  maf: NA
  probes: NA
  
  #exploratory analysis
  plot_violin_raw: NULL
  plot_violin_normalized: NULL
  plot_pca: NULL
  table_age: NULL
  plot_random_heatmap: NULL
  plot_top_heatmap: NULL
  plot_deconvolution: NULL
  plot_hyper_hypo_chr: NULL
  plot_hyper_hypo_relation_to_island: NULL
  plot_hyper_hypo_group: NULL
  
  #DMP 
  limma_voi: NA
  limma_covar: NA
  limma_inter: NA
  limma_arrayweights: NA
  limma_ebayes_trend: NA
  limma_ebayes_robust: NA
  

  rval_design: NA
  rval_contrasts: NA
  rval_voi: NA
  rval_dendrogram: NULL
  
  table_annotation_manhattan: NULL
  plot_volcano: NULL
  table_annotation: NULL
  
  
  #heatmap 
  min_deltabeta: NA
  max_fdr: NA
  max_pvalue: NA
  clusteralg: NA
  groups2plot: NA
  contrasts2plot: NA
  Colv: NA
  distance: NA
  scale: NA
  removebatch: NA
  
  #DMRs
  dmrs_contrasts: NULL
  dmrs_rval_dendrogram: NULL
  dmrs_min_deltabeta: NA
  dmrs_max_fdr: NA
  dmrs_max_pvalue: NA
  dmrs_clusteralg: NA
  dmrs_groups2plot: NA
  dmrs_contrasts2plot: NA
  dmrs_regions2plot: NA
  dmrs_Colv: NA
  dmrs_distance: NA
  dmrs_scale: NA
  dmrs_removebatch: NA
  table_dmrs: NA
  filteredmcsea2heatmap: NULL
  table_sigdmrs: NULL

  
  #plots

  plot_green_intensities: NA
  plot_red_intensities: NA
  plot_failed_probes: NA
  plot_densityplotraw: NA
  plot_densityplotraw_green: NA
  plot_densityplotraw_red: NA
  plot_densityplotraw_II: NA
  plot_densityplotraw_all: NA
  plot_densityplot: NA
  plot_densityplot_green: NA
  plot_densityplot_red: NA
  plot_densityplot_II: NA
  plot_densityplot_all: NA
  plot_pcaplot: NA
  plot_corrplot: NA
  plot_boxplotraw: NA
  plot_boxplot: NA
  plot_qcraw: NA
  plot_bisulfiterawII: NA
  plot_sexprediction: NA
  plot_snpheatmap: NA
  plot_plotSA: NA
  
  table_pcaplot: NA 
  table_corrplot: NA 
  data_sexprediction: NA
  
  plot_heatmap_html: NA
  table_dmps: NA
  filteredlist2heatmap: NULL
  
  
  
  #functional enrichment
  kegg: NULL
  go_mf: NULL
  go_bp: NULL
  go_cc: NULL
  reactome: NULL

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{css, echo=FALSE}
div, code {overflow-x:auto !important}
```

## Data

The group variable selected was __`r params$grouping_var`__, and the donor variable was __`r params$donor_var`__ .

Finally, the selected samples added to the RGSet file were:

```{r sample_sheet, echo=FALSE, warning=FALSE, message = FALSE}
DT::datatable(params$rval_sheet_target, extensions = 'Buttons',
    style = "bootstrap",
    rownames = FALSE,
    selection = "single",
    options = list(dom = 'Blfrtip',
    columnDefs = list(list(
      targets = match("Basename",colnames(params$rval_sheet_target)) - 1, visible = FALSE)),
     #scrollX = TRUE,
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10,25,50,-1), c(10,25,50,"All"))))
```

The following samples were excluded for the analysis:

```{r samples_excluded, echo=FALSE, warning=FALSE, message = FALSE, comment=NA}

cat(params$rval_sheet[[params$name_var]][!(params$rval_sheet[[params$name_var]] %in% params$rval_sheet_target[[params$name_var]])], 
    sep = ", ")

```

***

## Quality Control

### Intensities boxplots

__Green intensities__

```{r green_intensities, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_green_intensities
```

__Red intensities__

```{r red_intensities, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_red_intensities
```

<br>

### Failed probes

The probe failure rate is calculated for each sample and two thresholds are represented (5%, 10%).

```{r failed_probes, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_failed_probes
```

<br>

### Normalization options

The selected normalization method was __`r params$normalization_mode`__. In the next plots, we can see the comparison of the processed data with the raw data.

After Minfi normalization, we obtained a GenomicRatioSet with some transformations depending on the type of normalization chosen. To understand better the different classes of Minfi packages, and their relations depending of the normalization method, please read this [vignette](https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html) of Minfi creators (it is also valid for Illumina EPIC arrays).

Options selected:

* Drop SNPs: __`r params$dropsnps`__ *`r if(params$dropsnps) paste('minimum MAF filtered:', params$maf)`*
* Drop CpHs: __`r params$dropcphs`__
* Drop X/Y Chromosomes: __`r params$dropsex`__

After normalization, __`r params$probes`__ DNA methylation positions remained.

<br>

#### Density plot  

__Raw__

```{r density_plot_raw_Igreen, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_densityplotraw_all
```

__Processed `r params$normalization_mode`__

```{r density_plot_processed_Igreen, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_densityplot_all
```

<br>

### SNPs Heatmap

The methylation array has 65 specific SNP probes. These SNP probes are intended to be used for sample tracking and sample mixups. Each SNP probe ought to have values clustered around 3 distinct values corresponding to homo-, and hetero-zygotes. Therefore, different samples of the same donor should cluster together.

__Processed `r params$normalization_mode`__

```{r snps_heatmap, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_snpheatmap
```

<br>

### Sex Prediction

Depending on the average chromosome Y signal and the average chromosome X signal is possible to predict the sex of the sample donors.

```{r sex_prediction_plot, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_sexprediction
```

```{r sex_prediction_table, echo=FALSE, warning=FALSE, message = FALSE}
knitr::kable(data.frame(name=params$rval_sheet_target[[params$name_var]],sex=params$data_sexprediction))
```

<br>

### Batch effects

Correlating principal components with variables we can determine if Beta values are related to our variable of interest or other variables. This can also be useful to determine possible errors in the sample hybridization randomization and to select covariables to add to the linear model.


```{r corrplot, echo=FALSE, warning=FALSE, message = FALSE}
params$plot_corrplot
```

Not useful variables are discarded and the variable type is autodetected. The autodetected variable types were:

```{r corrtable, echo=FALSE, warning=FALSE, message = FALSE}
DT::datatable(params$table_corrplot, extensions = 'Buttons',
    style = 'bootstrap',
    rownames = FALSE,
    selection = "single",
    options = list(dom = 'Blfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10,25,50,-1),
    c(10,25,50,"All"))))
```

***

## Exploratory Analysis

### Violin plot

```{r violin, echo = FALSE, warning = FALSE, message = FALSE}
params$plot_violin_raw
params$plot_violin_normalized
```

### Principal Component Analysis

``` {r pca, echo = FALSE, warnimg = FALSE, message = FALSE}
params$plot_pca
```

### Heatmaps

``` {r heatmaps_random, echo = FALSE, warning = FALSE, message = FALSE}
params$plot_random_heatmap
```
``` {r heatmaps_top, echo = FALSE, warning = FALSE, message = FALSE}
params$plot_top_heatmap
```

### Age methylation

```{r age_meth, echo = FALSE, warning =FALSE, message = FALSE}
DT::datatable(params$table_age)
```

### Deconvolution

```{r deconvolution, echo = FALSE, warning = FALSE, message = FALSE}
params$plot_deconvolution
```

### Hypo/Hyper plots

```{r hypo_hyper, echo = FALSE, warning = FALSE, message = FALSE}
params$plot_hyper_hypo_chr
params$plot_hyper_hypo_relation_to_island
params$plot_hyper_hypo_group
```

***

```{r linear_model, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\n## DMPs and DMRs")

cat("\n### Linear Model")

cat("\nThe linear model was generated using ", params$limma_voi, " as variable of interest, ", if(length(params$limma_covar) == 0) "None" else paste(params$limma_covar, collapse=", "), " as covariable(s), and ", if(length(params$limma_inter) == 0) "None" else paste(params$limma_inter, collapse=", "), " as interaction(s).") 

cat("\nThe resulting design matrix was:")
```

```{r design, echo=FALSE, warning=FALSE, message = FALSE, eval = !is.null(params$filteredlist2heatmap)}
if (!is.na(params$rval_design)){
DT::datatable(
  params$rval_design,
  selection = "single",
  rownames = TRUE,
  style = "bootstrap",
  options(list(
    #scrollX = TRUE,
    lengthChange = FALSE,
    searching = FALSE
  ))
)}
```

```{r array_weights, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nOptionally, ArrayWeights can be activated to weight the value of each sample depending on their adjust to the linear model:")
```

```{r limma, echo=!is.null(params$filteredlist2heatmap), eval=FALSE}
 
      if (weights_option){
        try({weights = limma::arrayWeights(MValues, design = design})
      }
        else { weights = NULL}
        
      fit = limma::lmFit(MValues, design, weights = weights)
```

```{r array_weights_option, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nWhere `weights` is the option specified in the form. In this case, weights is ", params$limma_arrayweights)
```

```{r contrasts, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nWith the sample groups selected, the possible contrasts were:\n")
cat(params$rval_contrasts, sep = "\n")
```

```{r contrasts_statistics, echo = FALSE, eval = !is.null(params$filteredlist2heatmap),results='asis'}
cat("\nUsing the calculated model, the t-statistics and p.values are obtained automatically for each contrast:")
```

```{r ebayes, eval=FALSE, echo=!is.null(params$filteredlist2heatmap)}
  mcontrast = makeContrasts(contrasts=contrast, levels = design)
  fitting = contrasts.fit(fit, mcontrast)
  fitting = eBayes(fitting, trend=trend, robust=robust)
  toptable = topTable(fitting, coef= 1,  adjust.method = "fdr", number =  Inf)
```

```{r ebayes_options, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nWhere `trend` and `robust` are the options specified in the form. In this case, trend is ", params$limma_ebayes_trend, " and robust is ", params$limma_ebayes_robust)
```

<br>

```{r dmp_options, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\n### DMPs")
cat("\n#### Options (DMPs)")
cat("\nFor each contrast (", paste(params$rval_contrasts, collapse=", "), " ), the statistics for each CpG were generated and, after, filtered with these criteria:\n")
cat("* Max. FDR: ", params$max_fdr,
"\n* Min. Δβ: ", params$min_deltabeta,
"\n* Max. pvalue: ", params$max_pvalue, "\n")

cat("\n#### Table with all contrasts (DMPs)")
```

```{r filtered_table_all,echo=FALSE,message=FALSE, warning=FALSE}
if(!is.null(params$filteredlist2heatmap)) knitr::kable(params$table_dmps)
```

```{r dmp_custom_heatmap, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\n#### Custom heatmap (DMPs)")
cat("\n* Column dendogram: ", params$Colv,
"\n* Scale: ", params$scale,
"\n* Cluster Algorithm: ", params$clusteralg,
"\n* Distance: ", params$distance)

cat("\nThe data used for the representation are ", if(params$removebatch == FALSE) "the normalized beta values" else "the normalized beta values after removing covariable effects on the data (with the limma removeBatchEffect function), preserving the variable of interest")

cat("\nIn the custom heatmap the selected contrasts to plot were: ", paste(params$contrasts2plot, collapse=", "), " and the groups were: ", paste(params$groups2plot, collapse=", "))
```

```{r heatmap_dmps,echo=FALSE,message=FALSE, warning=FALSE, fig.width = 9, fig.height = 12, eval = !is.null(params$filteredlist2heatmap)}
    if(!is.null(params$filteredlist2heatmap)){
      create_heatmap(params$filteredlist2heatmap, 
                     factorgroups = factor(params$rval_voi[params$rval_voi %in% params$groups2plot], levels = params$groups2plot), 
                     groups2plot = params$rval_voi %in% params$groups2plot, 
                     Colv = as.logical(params$Colv), 
                     clusteralg = params$clusteralg, 
                     distance = params$distance,
                     scale = params$scale, 
                     static=TRUE, 
                     ColSideColors = TRUE,
                     RowSideColors = params$rval_dendrogram) 
      } else {
        cat("Differences are not in the plotting range (<12000, >1)"
            )}
  
```

```{r num_dmps_heatmap, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nDMPs in heatmap: ", nrow(params$filteredlist2heatmap))
```

```{r manhattan, echo=FALSE, eval=!is.null(params$filteredlist2heatmap), results='asis'}
cat("\n#### Manhattan plot")
```
```{r plot_manhattan, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$filteredlist2heatmap)}
qqman::manhattan(params$table_annotation_manhattan, chr = "chr", bp = "pos", snp = "gene", p = "pvalue",
                                                 annotatePval = 1, suggestiveline = T, genomewideline = T, annotateTop = T)
```

```{r volcano, echo=FALSE, eval=!is.null(params$filteredlist2heatmap), results='asis'}
cat("\n#### Volcano plot")
```
```{r plot_volcano, echo=FALSE, message=FALSE, warning=FALSE, eval=!is.null(params$filteredlist2heatmap)}
params$plot_volcano
```

```{r annotation_dmps, echo=FALSE, eval=!is.null(params$filteredlist2heatmap), results='asis'}
cat("\n#### Annotation (DMPs)")
```
```{r table_annotation_dmps, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$filteredlist2heatmap)}
DT::datatable(params$table_annotation,
        extensions = "Buttons",
        rownames = FALSE,
        selection = "single",
        style = "bootstrap",
        caption = "DMPs Annotation:",
        options = list(
            dom = "Blfrtip",
            lengthMenu = list(c(10, 25, 50, 100, 25000), c(10, 25, 50, 100, 25000)),
            pageLength = 10,
            autoWidth = TRUE,
            scrollX = TRUE,
            buttons = c("csv", "excel", "print")
        ))
```

<br>

```{r filt_res, echo=FALSE, eval=!is.null(params$dmrs_contrasts), results="asis"}

cat("\n### DMRs")
cat("\n#### Options (DMRs)")
cat("\nFor each contrast (", paste(params$rval_contrasts, collapse=", "))
cat(" ) the statistics for each CpG were generated and, after, filtered with these criteria: \n
* Max. FDR: ", params$dmrs_max_fdr, "\n* Min. Δβ: ", params$dmrs_min_deltabeta, "\n* Max. pvalue: ", params$dmrs_max_pvalue, "\n")
cat("\n#### Table with all contrasts (DMRs)")
```

```{r filtered_table_all_dmrs,echo=FALSE,message=FALSE, warning=FALSE}
if(!is.null(params$dmrs_contrasts)) knitr::kable(params$table_dmrs)
```


```{r custom_heatmap, echo=FALSE, eval=!is.null(params$dmrs_contrasts), results="asis"}
cat("\n#### Custom heatmap (DMRs)")
cat("\n* Column dendogram:", params$dmrs_Colv,
"\n* Scale:", params$dmrs_scale,
"\n* Cluster Algorithm:", params$dmrs_clusteralg,
"\n* Distance:", params$dmrs_distance,
"\nThe data used for the representation are ", if(!params$dmrs_removebatch) "the normalized beta values" else "the normalized beta values after removing covariable effects on the data (with the limma removeBatchEffect function), preserving the variable of interest",
"\nIn the custom heatmap the selected contrasts to plot were: ", paste(params$dmrs_contrasts2plot, collapse=", "),
" and the groups were: ", paste(params$dmrs_groups2plot, collapse=", "))
```

```{r heatmap_dmrs,echo=FALSE,message=FALSE, warning=FALSE, fig.width = 9, fig.height = 12}
if(!is.null(params$filteredmcsea2heatmap) & !is.null(params$dmrs_contrasts)){
  create_heatmap(params$filteredmcsea2heatmap,
                 factorgroups =  factor(params$rval_voi[params$rval_voi %in% params$dmrs_groups2plot], levels = params$dmrs_groups2plot),
                 groups2plot = params$rval_voi %in% params$dmrs_groups2plot,
                 Colv = as.logical(params$dmrs_Colv),
                 clusteralg = params$dmrs_clusteralg,
                 distance = params$dmrs_distance,
                 scale = params$dmrs_scale,
                 static=TRUE,
                 ColSideColors = TRUE,
                 RowSideColors = params$dmrs_rval_dendrogram)
  }
  
```

```{r num_heatmap_dmrs, echo = FALSE, eval = !is.null(params$filteredmcsea2heatmap), results = "asis"}
cat("\nDMRs in heatmap: ", nrow(params$filteredmcsea2heatmap))
```

```{r annotation_dmrs, echo=FALSE, eval=!is.null(params$filteredmcsea2heatmap), results='asis'}
cat("\n#### Annotation (DMRs)")
```
```{r table_annotation_dmrs, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$filteredmcsea2heatmap)}
DT::datatable(params$table_sigdmrs,
        rownames = TRUE,
        extensions = "Buttons",
        selection = "single",
        style = "bootstrap",
        caption = "Select DMR to plot:",
        options = list(
            pageLength = 10,
            dom = "Blfrtip",
            lengthMenu = list(c(10, 25, 50, 100, 25000), c(10, 25, 50, 100, 25000)),
            autoWidth = TRUE,
            scrollX = TRUE,
            buttons = c("csv", "excel", "print")
        )
    )
```

***

```{r functional_enrichment, echo = FALSE, eval = !is.null(params$kegg), results = "asis"}
cat("\n## Functional enrichment")
cat("\n### KEGG")
```
```{r plot_kegg, echo = FALSE, warning = FALSE, message = FALSE, eval = !is.null(params$kegg), fig.width=15, fig.height=10}
params$kegg
```

```{r mf, echo = FALSE, eval = !is.null(params$go_mf), results='asis'}
cat("\n### GO - Molecular Function")
```
```{r plot_mf, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$go_mf), fig.width=15, fig.height=10}
params$go_mf
```

```{r bp, echo=FALSE, eval=!is.null(params$go_bp), results='asis'}
cat("\n### GO - Biological Process")
```
```{r plot_bp, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$go_bp), fig.width=15, fig.height=10}
params$go_bp
```

```{r cc, echo=FALSE, eval=!is.null(params$go_cc), results='asis'}
cat("\n### GO - Cellular Component")
```
```{r plot_cc, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$go_cc), fig.width=15, fig.height=10}
params$go_cc
```

```{r reactome, echo=FALSE, eval=!is.null(params$reactome), results='asis'}
cat("\n### Reactome")
```
```{r plot_reactome, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$reactome), fig.width=15, fig.height=10}
params$reactome
```

***

## System and analysis information

This report was generated at __`r Sys.time()`__.

The session information was the following:

```{r system_info, echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
print(sessionInfo())
```






