---
title: "Complete Report"
output: 
  html_document:
    toc: true
    toc_float:
        collapsed: true
  theme: sandstone
  highlight: pygments
toc-title: "Table of contents"
params:
  
  #checks
  check_qc: FALSE
  check_exploratory_analysis: FALSE
  
  #input 
  rval_sheet: NULL
  rval_sheet_target: NULL
  name_var: NULL
  grouping_var: NULL
  donor_var: NULL
  
  #normalization 
  normalization_mode: NULL
  dropsnps: NULL
  dropcphs: NULL
  dropsex: NULL
  maf: NULL
  probes: NULL
  
  #exploratory analysis
  plot_violin_raw: NULL
  plot_violin_normalized: NULL
  plot_pca: NULL
  table_age: NULL
  plot_random_heatmap: NULL
  plot_top_heatmap: NULL
  plot_deconvolution: NULL
  plot_hyper_hypo_chr: NULL
  plot_hyper_hypo_relation_to_island: NULL
  plot_hyper_hypo_group: NULL
  
  #DMP 
  limma_voi: NULL
  limma_covar: NULL
  limma_inter: NULL
  limma_arrayweights: NULL
  limma_ebayes_trend: NULL
  limma_ebayes_robust: NULL
  
  rval_design: NULL
  rval_contrasts: NULL
  rval_voi: NULL
  rval_dendrogram: NULL
  
  table_annotation_manhattan: NULL
  plot_volcano: NULL
  table_annotation: NULL
  
  #heatmap 
  min_deltabeta: NULL
  max_fdr: NULL
  max_pvalue: NULL
  clusteralg: NULL
  groups2plot: NULL
  contrasts2plot: NULL
  Colv: NULL
  distance: NULL
  scale: NULL
  removebatch: NULL
  
  #DMRs
  dmrs_contrasts: NULL
  dmrs_rval_dendrogram: NULL
  dmrs_min_deltabeta: NULL
  dmrs_max_fdr: NULL
  dmrs_max_pvalue: NULL
  dmrs_clusteralg: NULL
  dmrs_groups2plot: NULL
  dmrs_contrasts2plot: NULL
  dmrs_regions2plot: NULL
  dmrs_Colv: NULL
  dmrs_distance: NULL
  dmrs_scale: NULL
  dmrs_removebatch: NULL
  table_dmrs: NULL
  filteredmcsea2heatmap: NULL
  table_sigdmrs: NULL

  
  #plots

  plot_green_intensities: NULL
  plot_red_intensities: NULL
  plot_failed_probes: NULL
  plot_densityplotraw: NULL
  plot_densityplotraw_green: NULL
  plot_densityplotraw_red: NULL
  plot_densityplotraw_II: NULL
  plot_densityplotraw_all: NULL
  plot_densityplot: NULL
  plot_densityplot_green: NULL
  plot_densityplot_red: NULL
  plot_densityplot_II: NULL
  plot_densityplot_all: NULL
  plot_pcaplot: NULL
  plot_corrplot: NULL
  plot_boxplotraw: NULL
  plot_boxplot: NULL
  plot_qcraw: NULL
  plot_bisulfiterawII: NULL
  plot_sexprediction: NULL
  plot_snpheatmap: NULL
  plot_plotSA: NULL
  
  table_pcaplot: NULL 
  table_corrplot: NULL 
  data_sexprediction: NULL
  
  plot_heatmap_html: NULL
  table_dmps: NULL
  filteredlist2heatmap: NULL
  
  #functional enrichment
  kegg: NULL
  go_mf: NULL
  go_bp: NULL
  go_cc: NULL
  reactome: NULL

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{css, echo=FALSE}
div, code {overflow-x:auto !important}
```

## Data

The group variable selected was __`r params$grouping_var`__, and the donor variable was __`r params$donor_var`__ .

Finally, the selected samples added to the RGSet file were:

```{r sample_sheet, echo=FALSE, warning=FALSE, message = FALSE}
DT::datatable(params$rval_sheet_target, extensions = 'Buttons',
    style = "bootstrap",
    rownames = FALSE,
    selection = "single",
    options = list(dom = 'Blfrtip',
    columnDefs = list(list(
      targets = match("Basename",colnames(params$rval_sheet_target)) - 1, visible = FALSE)),
     #scrollX = TRUE,
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10,25,50,-1), c(10,25,50,"All"))))
```

The following samples were excluded for the analysis:

```{r samples_excluded, echo=FALSE, warning=FALSE, message = FALSE, comment=NA}

cat(params$rval_sheet[[params$name_var]][!(params$rval_sheet[[params$name_var]] %in% params$rval_sheet_target[[params$name_var]])], 
    sep = ", ")

```

***

```{r quality_control, echo = FALSE, eval=params$check_qc==TRUE, results='asis'}
cat("\n## Quality Control")
```

```{r green_intensities, echo=FALSE, eval = !is.null(params$plot_green_intensities), results='asis'}
cat("\n### Intensities boxplots")

cat("\n")
cat("####\n")
cat("\n")

cat("\n__Green intensities__")

```

```{r plot_green_intensities, echo=FALSE, warning=FALSE, message = FALSE, eval = !is.null(params$plot_green_intensities)}
params$plot_green_intensities
```

```{r red_intensities, echo=FALSE, eval=!is.null(params$plot_red_intensities), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n__Red intensities__")
```
```{r plot_red_intensities, echo=FALSE, warning=FALSE, message = FALSE, eval = !is.null(params$plot_red_intensities)}
params$plot_red_intensities
```

```{r failed_probes, echo = FALSE, eval = !is.null(params$plot_failed_probes), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Failed probes")
cat("\nThe probe failure rate is calculated for each sample and two thresholds are represented (5%, 10%).")
```
```{r plot_failed_probes, echo=FALSE, warning=FALSE, message = FALSE, eval = !is.null(params$plot_failed_probes)}
params$plot_failed_probes
```

```{r normalization_options, echo = FALSE, eval=params$check_qc==TRUE, results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Normalization options")

cat("\nThe selected normalization method was ", params$normalization_mode, ". In the next plots, we can see the comparison of the processed data with the raw data.")

cat("\n After Minfi normalization, we obtained a GenomicRatioSet with some transformations depending on the type of normalization chosen. To understand better the different classes of Minfi packages, and their relations depending of the normalization method, please read this [vignette](https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html) of Minfi creators (it is also valid for Illumina EPIC arrays).\n")

cat("\nOptions selected:\n\n")

cat("* Drop SNPs: ", params$dropsnps)
if(params$dropsnps){ cat(" minimum MAF filtered: ", params$maf)}
cat("\n* Drop CpHs: ", params$dropcphs,
"\n* Drop X/Y Chromosomes: ", params$dropsex)

cat("\n\nAfter normalization, ", params$probes, "DNA methylation positions remained.")
```

```{r density_raw, echo = FALSE, eval=!is.null(params$plot_densityplotraw_all), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n#### Density plot")  

cat("\n__Raw__")
```
```{r plot_density_raw, echo=FALSE, warning=FALSE, message = FALSE, eval=!is.null(params$plot_densityplotraw_all)}
params$plot_densityplotraw_all
```
```{r density_normalized, echo=FALSE, eval = !is.null(params$plot_densityplot_all), results='asis'}
cat("\n__Normalized ", params$normalization_mode, "__")
```
```{r plot_density_normalized, echo=FALSE, warning=FALSE, message = FALSE, eval = !is.null(params$plot_densityplot_all)}
params$plot_densityplot_all
```

```{r snps_heatmap, echo=FALSE, eval=!is.null(params$plot_snpheatmap), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### SNPs Heatmap")

cat("\nThe methylation array has 65 specific SNP probes. These SNP probes are intended to be used for sample tracking and sample mixups. Each SNP probe ought to have values clustered around 3 distinct values corresponding to homo-, and hetero-zygotes. Therefore, different samples of the same donor should cluster together.\n")
```
```{r plot_snps_heatmap, echo=FALSE, warning=FALSE, message = FALSE, eval=!is.null(params$plot_snpheatmap)}
params$plot_snpheatmap
```

```{r sex_prediction, echo=FALSE, eval=!is.null(params$plot_sexprediction), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Sex Prediction")

cat("\nDepending on the average chromosome Y signal and the average chromosome X signal is possible to predict the sex of the sample donors.")
```
```{r plot_sex_prediction, echo=FALSE, warning=FALSE, message = FALSE, eval=!is.null(params$plot_sexprediction)}
params$plot_sexprediction
```
```{r tale_sex_prediction, echo=FALSE, warning=FALSE, message = FALSE, eval=!is.null(params$data_sexprediction)}
knitr::kable(data.frame(name=params$rval_sheet_target[[params$name_var]],sex=params$data_sexprediction))
```

```{r batch_effects, echo=FALSE, eval=!is.null(params$plot_corrplot), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Batch effects")

cat("\nCorrelating principal components with variables we can determine if Beta values are related to our variable of interest or other variables. This can also be useful to determine possible errors in the sample hybridization randomization and to select covariables to add to the linear model.")
```
```{r plot_batch_effects, echo=FALSE, warning=FALSE, message = FALSE, eval=!is.null(params$plot_corrplot)}
params$plot_corrplot
```
```{r batch_efects_annotation, echo=FALSE, eval=!is.null(params$table_corrplot), results='asis'}
cat("\nNot useful variables are discarded and the variable type is autodetected. The autodetected variable types were:")
```
```{r table_batch_effects, echo=FALSE, warning=FALSE, message = FALSE, eval=!is.null(params$table_corrplot)}
DT::datatable(params$table_corrplot, extensions = 'Buttons',
    style = 'bootstrap',
    rownames = FALSE,
    selection = "single",
    options = list(dom = 'Blfrtip',
    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
    lengthMenu = list(c(10,25,50,-1),
    c(10,25,50,"All"))))
```

***

```{r exploratory_analysis, echo=FALSE, eval=params$check_exploratory_analysis==TRUE, results='asis'}
cat("\n## Exploratory Analysis")
```

```{r violin_raw, echo=FALSE, eval=!is.null(params$plot_violin_raw), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Violin plot\n")

cat("\n__Raw__\n")

cat("\n")
```
```{r plot_violin_raw, echo = FALSE, warning = FALSE, message = FALSE, eval=!is.null(params$plot_violin_raw)}
params$plot_violin_raw
```
```{r violin_normalized, echo=FALSE, eval=!is.null(params$plot_violin_normalized), results='asis'}
cat("\n__Normalized__\n")

cat("\n")
```
```{r plot_violin_normalized, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$plot_violin_normalized)}
params$plot_violin_normalized
```

```{r pca, echo=FALSE, eval=!is.null(params$plot_pca), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Principal Component Analysis\n")
```
```{r plot_pca, echo = FALSE, warnimg = FALSE, message = FALSE, eval=!is.null(params$plot_pca)}
params$plot_pca
```

```{r heatmap_random, echo=FALSE, eval=!is.null(params$plot_random_heatmap), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Heatmaps\n")

cat("\n__Random 1000 CpGs__\n")
```
```{r plot_heatmap_random, echo = FALSE, warning = FALSE, message = FALSE, eval=!is.null(params$plot_random_heatmap)}
params$plot_random_heatmap
```
```{r heatmap_top, echo=FALSE, eval=!is.null(params$plot_top_heatmap), results='asis'}
cat("\n__Top 1000 variable CpGs__\n")
```
```{r plot_heatmap_top, echo = FALSE, warning = FALSE, message = FALSE, eval=!is.null(params$plot_top_heatmap)}
params$plot_top_heatmap
```

```{r age_meth, echo=FALSE, eval=!is.null(params$table_age), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Age methylation\n")
```
```{r table_age_meth, echo = FALSE, warning =FALSE, message = FALSE, eval=!is.null(params$table_age)}
DT::datatable(params$table_age)
```

```{r deconvolution, echo=FALSE, eval=!is.null(params$plot_deconvolution), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Deconvolution\n")
```
```{r plot_deconvolution, echo = FALSE, warning = FALSE, message = FALSE, eval=!is.null(params$plot_deconvolution)}
params$plot_deconvolution
```

```{r hypo_hyper, echo=FALSE, eval=!is.null(params$plot_hyper_hypo_chr), results='asis'}
cat("\n")
cat("#####\n")
cat("\n")

cat("\n### Hypo/Hyper plots\n")
```
```{r plot_hypo_hyper, echo = FALSE, warning = FALSE, message = FALSE, eval=!is.null(params$plot_hyper_hypo_chr)}
params$plot_hyper_hypo_chr
params$plot_hyper_hypo_relation_to_island
params$plot_hyper_hypo_group
```

***

```{r linear_model, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\n## DMPs and DMRs")

cat("\n### Linear Model")

cat("\nThe linear model was generated using ", params$limma_voi, " as variable of interest, ", if(length(params$limma_covar) == 0) "None" else paste(params$limma_covar, collapse=", "), " as covariable(s), and ", if(length(params$limma_inter) == 0) "None" else paste(params$limma_inter, collapse=", "), " as interaction(s).") 

cat("\nThe resulting design matrix was:")
```

```{r design, echo=FALSE, warning=FALSE, message = FALSE, eval = !is.null(params$filteredlist2heatmap)}
if (!is.na(params$rval_design)){
DT::datatable(
  params$rval_design,
  selection = "single",
  rownames = TRUE,
  style = "bootstrap",
  options(list(
    #scrollX = TRUE,
    lengthChange = FALSE,
    searching = FALSE
  ))
)}
```

```{r array_weights, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nOptionally, ArrayWeights can be activated to weight the value of each sample depending on their adjust to the linear model:")
```

```{r limma, echo=!is.null(params$filteredlist2heatmap), eval=FALSE}
 
      if (weights_option){
        try({weights = limma::arrayWeights(MValues, design = design})
      }
        else { weights = NULL}
        
      fit = limma::lmFit(MValues, design, weights = weights)
```

```{r array_weights_option, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nWhere `weights` is the option specified in the form. In this case, weights is ", params$limma_arrayweights)
```

```{r contrasts, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nWith the sample groups selected, the possible contrasts were:\n")
cat(params$rval_contrasts, sep = "\n")
```

```{r contrasts_statistics, echo = FALSE, eval = !is.null(params$filteredlist2heatmap),results='asis'}
cat("\nUsing the calculated model, the t-statistics and p.values are obtained automatically for each contrast:")
```

```{r ebayes, eval=FALSE, echo=!is.null(params$filteredlist2heatmap)}
  mcontrast = makeContrasts(contrasts=contrast, levels = design)
  fitting = contrasts.fit(fit, mcontrast)
  fitting = eBayes(fitting, trend=trend, robust=robust)
  toptable = topTable(fitting, coef= 1,  adjust.method = "fdr", number =  Inf)
```

```{r ebayes_options, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nWhere `trend` and `robust` are the options specified in the form. In this case, trend is ", params$limma_ebayes_trend, " and robust is ", params$limma_ebayes_robust)
```

<br>

```{r dmp_options, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\n### DMPs")
cat("\n#### Options (DMPs)")
cat("\nFor each contrast (", paste(params$rval_contrasts, collapse=", "), " ), the statistics for each CpG were generated and, after, filtered with these criteria:\n")
cat("* Max. FDR: ", params$max_fdr,
"\n* Min. Δβ: ", params$min_deltabeta,
"\n* Max. pvalue: ", params$max_pvalue, "\n")

cat("\n#### Table with all contrasts (DMPs)")
```

```{r filtered_table_all,echo=FALSE,message=FALSE, warning=FALSE}
if(!is.null(params$filteredlist2heatmap)) knitr::kable(params$table_dmps)
```

```{r dmp_custom_heatmap, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\n#### Custom heatmap (DMPs)")
cat("\n* Column dendogram: ", params$Colv,
"\n* Scale: ", params$scale,
"\n* Cluster Algorithm: ", params$clusteralg,
"\n* Distance: ", params$distance)

cat("\nThe data used for the representation are ", if(params$removebatch == FALSE) "the normalized beta values" else "the normalized beta values after removing covariable effects on the data (with the limma removeBatchEffect function), preserving the variable of interest")

cat("\nIn the custom heatmap the selected contrasts to plot were: ", paste(params$contrasts2plot, collapse=", "), " and the groups were: ", paste(params$groups2plot, collapse=", "))
```

```{r heatmap_dmps,echo=FALSE,message=FALSE, warning=FALSE, fig.width = 9, fig.height = 12, eval = !is.null(params$filteredlist2heatmap)}
    if(!is.null(params$filteredlist2heatmap)){
      create_heatmap(params$filteredlist2heatmap, 
                     factorgroups = factor(params$rval_voi[params$rval_voi %in% params$groups2plot], levels = params$groups2plot), 
                     groups2plot = params$rval_voi %in% params$groups2plot, 
                     Colv = as.logical(params$Colv), 
                     clusteralg = params$clusteralg, 
                     distance = params$distance,
                     scale = params$scale, 
                     static=TRUE, 
                     ColSideColors = TRUE,
                     RowSideColors = params$rval_dendrogram) 
      } else {
        cat("Differences are not in the plotting range (<12000, >1)"
            )}
```

```{r num_dmps_heatmap, echo = FALSE, eval = !is.null(params$filteredlist2heatmap), results = "asis"}
cat("\nDMPs in heatmap: ", nrow(params$filteredlist2heatmap))
```

```{r manhattan, echo=FALSE, eval=!is.null(params$filteredlist2heatmap), results='asis'}
cat("\n#### Manhattan plot")
```
```{r plot_manhattan, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$filteredlist2heatmap)}
qqman::manhattan(params$table_annotation_manhattan, chr = "chr", bp = "pos", snp = "gene", p = "pvalue",
                                                 annotatePval = 1, suggestiveline = T, genomewideline = T, annotateTop = T)
```

```{r volcano, echo=FALSE, eval=!is.null(params$filteredlist2heatmap), results='asis'}
cat("\n#### Volcano plot")
```
```{r plot_volcano, echo=FALSE, message=FALSE, warning=FALSE, eval=!is.null(params$filteredlist2heatmap)}
params$plot_volcano
```

```{r annotation_dmps, echo=FALSE, eval=!is.null(params$filteredlist2heatmap), results='asis'}
cat("\n#### Annotation (DMPs)")
```
```{r table_annotation_dmps, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$filteredlist2heatmap)}
DT::datatable(params$table_annotation,
        extensions = "Buttons",
        rownames = FALSE,
        selection = "single",
        style = "bootstrap",
        caption = "DMPs Annotation:",
        options = list(
            dom = "Blfrtip",
            lengthMenu = list(c(10, 25, 50, 100, 25000), c(10, 25, 50, 100, 25000)),
            pageLength = 10,
            autoWidth = TRUE,
            scrollX = TRUE,
            buttons = c("csv", "excel", "print")
        ))
```

<br>

```{r filt_res, echo=FALSE, eval=!is.null(params$dmrs_contrasts), results="asis"}
cat("\n### DMRs")
cat("\n#### Options (DMRs)")
cat("\nFor each contrast (", paste(params$rval_contrasts, collapse=", "))
cat(" ) the statistics for each CpG were generated and, after, filtered with these criteria: \n
* Max. FDR: ", params$dmrs_max_fdr, "\n* Min. Δβ: ", params$dmrs_min_deltabeta, "\n* Max. pvalue: ", params$dmrs_max_pvalue, "\n")
cat("\n#### Table with all contrasts (DMRs)")
```

```{r filtered_table_all_dmrs,echo=FALSE,message=FALSE, warning=FALSE}
if(!is.null(params$dmrs_contrasts)) knitr::kable(params$table_dmrs)
```


```{r custom_heatmap, echo=FALSE, eval=!is.null(params$dmrs_contrasts), results="asis"}
cat("\n#### Custom heatmap (DMRs)")
cat("\n* Column dendogram:", params$dmrs_Colv,
"\n* Scale:", params$dmrs_scale,
"\n* Cluster Algorithm:", params$dmrs_clusteralg,
"\n* Distance:", params$dmrs_distance,
"\nThe data used for the representation are ", if(!params$dmrs_removebatch) "the normalized beta values" else "the normalized beta values after removing covariable effects on the data (with the limma removeBatchEffect function), preserving the variable of interest",
"\nIn the custom heatmap the selected contrasts to plot were: ", paste(params$dmrs_contrasts2plot, collapse=", "),
" and the groups were: ", paste(params$dmrs_groups2plot, collapse=", "))
```

```{r heatmap_dmrs,echo=FALSE,message=FALSE, warning=FALSE, fig.width = 9, fig.height = 12}
if(!is.null(params$filteredmcsea2heatmap) & !is.null(params$dmrs_contrasts)){
  create_heatmap(params$filteredmcsea2heatmap,
                 factorgroups =  factor(params$rval_voi[params$rval_voi %in% params$dmrs_groups2plot], levels = params$dmrs_groups2plot),
                 groups2plot = params$rval_voi %in% params$dmrs_groups2plot,
                 Colv = as.logical(params$dmrs_Colv),
                 clusteralg = params$dmrs_clusteralg,
                 distance = params$dmrs_distance,
                 scale = params$dmrs_scale,
                 static=TRUE,
                 ColSideColors = TRUE,
                 RowSideColors = params$dmrs_rval_dendrogram)
  }
  
```

```{r num_heatmap_dmrs, echo = FALSE, eval = !is.null(params$filteredmcsea2heatmap), results = "asis"}
cat("\nDMRs in heatmap: ", nrow(params$filteredmcsea2heatmap))
```

```{r annotation_dmrs, echo=FALSE, eval=!is.null(params$filteredmcsea2heatmap), results='asis'}
cat("\n#### Annotation (DMRs)")
```
```{r table_annotation_dmrs, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$filteredmcsea2heatmap)}
DT::datatable(params$table_sigdmrs,
        rownames = TRUE,
        extensions = "Buttons",
        selection = "single",
        style = "bootstrap",
        caption = "Select DMR to plot:",
        options = list(
            pageLength = 10,
            dom = "Blfrtip",
            lengthMenu = list(c(10, 25, 50, 100, 25000), c(10, 25, 50, 100, 25000)),
            autoWidth = TRUE,
            scrollX = TRUE,
            buttons = c("csv", "excel", "print")
        )
    )
```

***

```{r functional_enrichment, echo = FALSE, eval = !is.null(params$kegg), results = "asis"}
cat("\n## Functional enrichment")
cat("\n### KEGG")
```
```{r plot_kegg, echo = FALSE, warning = FALSE, message = FALSE, eval = !is.null(params$kegg), fig.width=15, fig.height=10}
params$kegg
```

```{r mf, echo = FALSE, eval = !is.null(params$go_mf), results='asis'}
cat("\n### GO - Molecular Function")
```
```{r plot_mf, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$go_mf), fig.width=15, fig.height=10}
params$go_mf
```

```{r bp, echo=FALSE, eval=!is.null(params$go_bp), results='asis'}
cat("\n### GO - Biological Process")
```
```{r plot_bp, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$go_bp), fig.width=15, fig.height=10}
params$go_bp
```

```{r cc, echo=FALSE, eval=!is.null(params$go_cc), results='asis'}
cat("\n### GO - Cellular Component")
```
```{r plot_cc, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$go_cc), fig.width=15, fig.height=10}
params$go_cc
```

```{r reactome, echo=FALSE, eval=!is.null(params$reactome), results='asis'}
cat("\n### Reactome")
```
```{r plot_reactome, echo=FALSE, warning=FALSE, message=FALSE, eval=!is.null(params$reactome), fig.width=15, fig.height=10}
params$reactome
```

***

## System and analysis information

This report was generated at __`r Sys.time()`__.

The session information was the following:

```{r system_info, echo=FALSE, message=FALSE, warning=FALSE, comment=NA}
print(sessionInfo())
```






